File: apps/api/app/defaults.py
from __future__ import annotations

from .domain_types import (
    CostItem,
    Employee,
    ModelConfig,
    MONTH_LABELS,
    MonthlyPlan,
    MonthlyPlanTemplate,
    OperatingConfig,
    PlanningConfig,
    ScenarioBand,
    Shareholder,
    StageCostItem,
    StageCostMode,
    StageCostValue,
    TeamMember,
)


def create_stage_cost_item(seed: str, *, name: str, mode: StageCostMode) -> StageCostItem:
    return StageCostItem(id=f"stage-cost-{seed}", name=name, mode=mode)


def create_default_stage_cost_items() -> list[StageCostItem]:
    return [
        create_stage_cost_item("vj", name="舞台视觉", mode="monthly"),
        create_stage_cost_item("original-song", name="原创", mode="monthly"),
        create_stage_cost_item("makeup", name="化妆", mode="perEvent"),
        create_stage_cost_item("streaming", name="推流", mode="perEvent"),
        create_stage_cost_item("meal", name="聚餐", mode="perEvent"),
        create_stage_cost_item("team-building", name="团建", mode="perEvent"),
        create_stage_cost_item("material", name="耗材", mode="perUnit"),
    ]


def create_stage_cost_values(items: list[StageCostItem], values: list[StageCostValue] | None = None) -> list[StageCostValue]:
    value_map = {value.itemId: value for value in values or []}
    normalized: list[StageCostValue] = []
    for item in items:
        current = value_map.get(item.id)
        normalized.append(
            StageCostValue(
                itemId=item.id,
                amount=current.amount if current else 0,
                count=current.count if current else (0 if item.mode == "perEvent" else 1),
            )
        )
    return normalized


def get_month_label(start_month: int, index: int) -> str:
    normalized_start = min(12, max(1, round(start_month)))
    offset = (normalized_start - 1 + index) % 12
    return MONTH_LABELS[offset]


def create_default_model() -> ModelConfig:
    stage_cost_items = create_default_stage_cost_items()
    planning = PlanningConfig(startMonth=3, horizonMonths=12)
    startup_units = 0.66
    stable_units = 1.32
    timeline_template = MonthlyPlanTemplate(
        events=6,
        salesMultiplier=stable_units,
        onlineSalesFactor=0,
        rehearsalCount=4,
        rehearsalCost=300,
        teacherCount=4,
        teacherCost=200,
        specialCosts=create_stage_cost_values(
            stage_cost_items,
            [StageCostValue(itemId="stage-cost-material", amount=6, count=1)],
        ),
    )

    months: list[MonthlyPlan] = []
    for month_index in range(planning.horizonMonths):
        is_startup = month_index == 0
        stage_material_amount = 0 if month_index < 2 else 6
        months.append(
            MonthlyPlan(
                id=f"month-default-{month_index + 1}",
                label=get_month_label(planning.startMonth, month_index),
                events=6,
                salesMultiplier=startup_units if is_startup else stable_units,
                onlineSalesFactor=0,
                rehearsalCount=8 if is_startup else 4,
                rehearsalCost=300,
                teacherCount=8 if is_startup else 4,
                teacherCost=200,
                specialCosts=create_stage_cost_values(
                    stage_cost_items,
                    [StageCostValue(itemId="stage-cost-material", amount=stage_material_amount, count=1)],
                ),
            )
        )

    return ModelConfig(
        shareholders=[
            Shareholder(id="shareholder-a", name="股东 A", investmentAmount=65000, dividendRate=0.684),
            Shareholder(id="shareholder-b", name="股东 B", investmentAmount=30000, dividendRate=0.316),
        ],
        operating=OperatingConfig(
            offlineUnitPrice=88,
            onlineUnitPrice=88,
            monthlyFixedCosts=[],
            perEventCosts=[],
            perUnitCosts=[],
        ),
        planning=planning,
        stageCostItems=stage_cost_items,
        timelineTemplate=timeline_template,
        teamMembers=[
            TeamMember(
                id="member-lead",
                name="主成员",
                employmentType="salary",
                monthlyBasePay=1500,
                perEventTravelCost=0,
                departureMonthIndex=None,
                commissionRate=0.15,
                unitsPerEvent=ScenarioBand(pessimistic=20, base=32, optimistic=42),
            ),
            TeamMember(
                id="member-a",
                name="成员 A",
                employmentType="partTime",
                monthlyBasePay=0,
                perEventTravelCost=0,
                departureMonthIndex=None,
                commissionRate=0.35,
                unitsPerEvent=ScenarioBand(pessimistic=16, base=26, optimistic=32),
            ),
            TeamMember(
                id="member-b",
                name="成员 B",
                employmentType="partTime",
                monthlyBasePay=0,
                perEventTravelCost=0,
                departureMonthIndex=None,
                commissionRate=0.35,
                unitsPerEvent=ScenarioBand(pessimistic=15, base=23, optimistic=28),
            ),
            TeamMember(
                id="member-c",
                name="成员 C",
                employmentType="partTime",
                monthlyBasePay=0,
                perEventTravelCost=0,
                departureMonthIndex=None,
                commissionRate=0.35,
                unitsPerEvent=ScenarioBand(pessimistic=14, base=21, optimistic=26),
            ),
            TeamMember(
                id="member-d",
                name="成员 D",
                employmentType="partTime",
                monthlyBasePay=0,
                perEventTravelCost=0,
                departureMonthIndex=None,
                commissionRate=0.35,
                unitsPerEvent=ScenarioBand(pessimistic=13, base=19, optimistic=24),
            ),
            TeamMember(
                id="member-e",
                name="成员 E",
                employmentType="partTime",
                monthlyBasePay=0,
                perEventTravelCost=0,
                departureMonthIndex=None,
                commissionRate=0.35,
                unitsPerEvent=ScenarioBand(pessimistic=12, base=17, optimistic=22),
            ),
            TeamMember(
                id="member-f",
                name="成员 F",
                employmentType="partTime",
                monthlyBasePay=0,
                perEventTravelCost=0,
                departureMonthIndex=None,
                commissionRate=0.35,
                unitsPerEvent=ScenarioBand(pessimistic=10, base=14, optimistic=18),
            ),
        ],
        employees=[
            Employee(id="employee-a", name="员工 A", role="场务", monthlyBasePay=0, perEventCost=200),
            Employee(id="employee-b", name="员工 B", role="场务", monthlyBasePay=0, perEventCost=200),
        ],
        months=months,
    )

File: apps/web/src/lib/defaults.ts
import type {
  CostItem,
  Employee,
  ModelConfig,
  MonthlyPlan,
  MonthlyPlanTemplate,
  PlanningConfig,
  Shareholder,
  StageCostItem,
  StageCostMode,
  StageCostValue,
  TeamMember,
} from '../types'

export const monthLabelOptions = ['1月', '2月', '3月', '4月', '5月', '6月', '7月', '8月', '9月', '10月', '11月', '12月']

function createId(prefix: string, seed: string) {
  return `${prefix}-${seed}`
}

export function createShareholder(seed: string, values?: Partial<Shareholder>): Shareholder {
  return {
    id: createId('shareholder', seed),
    name: '股东',
    investmentAmount: 0,
    dividendRate: 0,
    ...values,
  }
}

export function createCostItem(seed: string, values?: Partial<CostItem>): CostItem {
  return {
    id: createId('cost', seed),
    name: '新成本项',
    amount: 0,
    ...values,
  }
}

export function createStageCostItem(seed: string, values?: Partial<StageCostItem>): StageCostItem {
  return {
    id: createId('stage-cost', seed),
    name: '新专项成本',
    mode: 'perEvent',
    ...values,
  }
}

export function createStageCostValue(
  itemId: string,
  mode: StageCostMode,
  values?: Partial<StageCostValue>,
): StageCostValue {
  return {
    itemId,
    amount: 0,
    count: mode === 'perEvent' ? 0 : 1,
    ...values,
  }
}

export function createDefaultStageCostItems() {
  return [
    createStageCostItem('vj', { name: '舞台视觉', mode: 'monthly' }),
    createStageCostItem('original-song', { name: '原创', mode: 'monthly' }),
    createStageCostItem('makeup', { name: '化妆', mode: 'perEvent' }),
    createStageCostItem('streaming', { name: '推流', mode: 'perEvent' }),
    createStageCostItem('meal', { name: '聚餐', mode: 'perEvent' }),
    createStageCostItem('team-building', { name: '团建', mode: 'perEvent' }),
    createStageCostItem('material', { name: '耗材', mode: 'perUnit' }),
  ] satisfies StageCostItem[]
}

export function createStageCostValues(items: StageCostItem[], values?: Partial<StageCostValue>[]) {
  const valueMap = new Map((values ?? []).map((item) => [item.itemId, item]))

  return items.map((item) =>
    createStageCostValue(item.id, item.mode, {
      ...(valueMap.get(item.id) ?? {}),
      itemId: item.id,
      count: item.mode === 'perEvent' ? valueMap.get(item.id)?.count ?? 0 : 1,
    }),
  )
}

export function createMember(seed: string, values?: Partial<TeamMember>): TeamMember {
  return {
    id: createId('member', seed),
    name: '新成员',
    employmentType: 'partTime',
    monthlyBasePay: 0,
    perEventTravelCost: 0,
    departureMonthIndex: null,
    commissionRate: 0.35,
    unitsPerEvent: {
      pessimistic: 8,
      base: 12,
      optimistic: 16,
    },
    ...values,
  }
}

export function createEmployee(seed: string, values?: Partial<Employee>): Employee {
  return {
    id: createId('employee', seed),
    name: '新员工',
    role: '现场执行',
    monthlyBasePay: 0,
    perEventCost: 200,
    ...values,
  }
}

export function createTimelineTemplate(
  values?: Partial<MonthlyPlanTemplate>,
  stageCostItems: StageCostItem[] = createDefaultStageCostItems(),
): MonthlyPlanTemplate {
  const nextTemplate = {
    events: 6,
    salesMultiplier: 1,
    onlineSalesFactor: 0,
    rehearsalCount: 4,
    rehearsalCost: 300,
    teacherCount: 4,
    teacherCost: 200,
    specialCosts: createStageCostValues(stageCostItems),
    ...values,
  }

  return {
    ...nextTemplate,
    specialCosts: createStageCostValues(stageCostItems, nextTemplate.specialCosts),
  }
}

export function toTimelineTemplate(
  month?: Partial<MonthlyPlan>,
  stageCostItems: StageCostItem[] = createDefaultStageCostItems(),
): MonthlyPlanTemplate {
  if (!month) {
    return createTimelineTemplate(undefined, stageCostItems)
  }

  const { id: _id, label: _label, ...templateValues } = month
  return createTimelineTemplate(templateValues, stageCostItems)
}

export function createMonth(
  seed: string,
  values?: Partial<MonthlyPlan>,
  stageCostItems: StageCostItem[] = createDefaultStageCostItems(),
): MonthlyPlan {
  const nextMonth = {
    id: createId('month', seed),
    label: '新月份',
    ...createTimelineTemplate(undefined, stageCostItems),
    ...values,
  }

  return {
    ...nextMonth,
    specialCosts: createStageCostValues(stageCostItems, nextMonth.specialCosts),
  }
}

export function getMonthLabel(startMonth: number, index: number) {
  const normalizedStart = Number.isFinite(startMonth) ? Math.min(12, Math.max(1, Math.round(startMonth))) : 1
  const offset = (normalizedStart - 1 + index) % 12

  return monthLabelOptions[offset] ?? `${normalizedStart}月`
}

function createMonthId(seed: string, index: number) {
  return createId('month', `${seed}-${index}-${Date.now()}`)
}

export function syncMonthsToPlanning(
  months: MonthlyPlan[],
  planning: PlanningConfig,
  seed = 'timeline',
  template: MonthlyPlanTemplate = createTimelineTemplate(),
  stageCostItems: StageCostItem[] = createDefaultStageCostItems(),
): MonthlyPlan[] {
  const horizonMonths = Number.isFinite(planning.horizonMonths)
    ? Math.min(24, Math.max(1, Math.round(planning.horizonMonths)))
    : 6
  const startMonth = Number.isFinite(planning.startMonth)
    ? Math.min(12, Math.max(1, Math.round(planning.startMonth)))
    : 1

  return Array.from({ length: horizonMonths }, (_, index) => {
    const current = months[index]

    if (current) {
      return {
        ...current,
        label: getMonthLabel(startMonth, index),
        specialCosts: createStageCostValues(stageCostItems, current.specialCosts),
      }
    }

    return createMonth(
      createMonthId(seed, index),
      {
        ...template,
        id: createMonthId(seed, index),
        label: getMonthLabel(startMonth, index),
      },
      stageCostItems,
    )
  })
}

export function createProductDefaultModel(): ModelConfig {
  const stageCostItems = createDefaultStageCostItems()
  const planning: PlanningConfig = {
    startMonth: 3,
    horizonMonths: 12,
  }
  const startupUnitsPerEvent = 0.66
  const stableUnitsPerEvent = 1.32

  const timelineTemplate = createTimelineTemplate(
    {
      events: 6,
      salesMultiplier: stableUnitsPerEvent,
      onlineSalesFactor: 0,
      rehearsalCount: 4,
      rehearsalCost: 300,
      teacherCount: 4,
      teacherCost: 200,
      specialCosts: [
        {
          itemId: 'stage-cost-material',
          amount: 6,
          count: 1,
        },
      ],
    },
    stageCostItems,
  )

  return {
    shareholders: [
      createShareholder('a', {
        name: '股东 A',
        investmentAmount: 65000,
        dividendRate: 0.684,
      }),
      createShareholder('b', {
        name: '股东 B',
        investmentAmount: 30000,
        dividendRate: 0.316,
      }),
    ],
    operating: {
      offlineUnitPrice: 88,
      onlineUnitPrice: 88,
      polaroidLossRate: 0.1,
      monthlyFixedCosts: [],
      perEventCosts: [],
      perUnitCosts: [],
    },
    planning,
    stageCostItems,
    timelineTemplate,
    teamMembers: [
      createMember('wenchen', {
        name: '文臣',
        employmentType: 'salary',
        monthlyBasePay: 1500,
        commissionRate: 0.15,
        unitsPerEvent: {
          pessimistic: 20,
          base: 32,
          optimistic: 42,
        },
      }),
      createMember('member-a', {
        name: '成员 A',
        unitsPerEvent: {
          pessimistic: 16,
          base: 26,
          optimistic: 32,
        },
      }),
      createMember('member-b', {
        name: '成员 B',
        unitsPerEvent: {
          pessimistic: 15,
          base: 23,
          optimistic: 28,
        },
      }),
      createMember('member-c', {
        name: '成员 C',
        unitsPerEvent: {
          pessimistic: 14,
          base: 21,
          optimistic: 26,
        },
      }),
      createMember('member-d', {
        name: '成员 D',
        unitsPerEvent: {
          pessimistic: 13,
          base: 19,
          optimistic: 24,
        },
      }),
      createMember('member-e', {
        name: '成员 E',
        unitsPerEvent: {
          pessimistic: 12,
          base: 17,
          optimistic: 22,
        },
      }),
      createMember('member-f', {
        name: '成员 F',
        unitsPerEvent: {
          pessimistic: 10,
          base: 14,
          optimistic: 18,
        },
      }),
    ],
    employees: [
      createEmployee('staff-a', {
        name: '员工 A',
        role: '场务',
        perEventCost: 200,
      }),
      createEmployee('staff-b', {
        name: '员工 B',
        role: '场务',
        perEventCost: 200,
      }),
    ],
    months: syncMonthsToPlanning(
      [
        createMonth(
          'mar',
          {
            label: '3月',
            events: 6,
            salesMultiplier: startupUnitsPerEvent,
            onlineSalesFactor: 0,
            rehearsalCount: 8,
            rehearsalCost: 300,
            teacherCount: 8,
            teacherCost: 200,
            specialCosts: createStageCostValues(stageCostItems, [
              {
                itemId: 'stage-cost-material',
                amount: 0,
                count: 1,
              },
            ]),
          },
          stageCostItems,
        ),
        createMonth(
          'apr',
          {
            label: '4月',
            events: 6,
            salesMultiplier: stableUnitsPerEvent,
            onlineSalesFactor: 0,
            rehearsalCount: 4,
            rehearsalCost: 300,
            teacherCount: 4,
            teacherCost: 200,
            specialCosts: createStageCostValues(stageCostItems, [
              {
                itemId: 'stage-cost-material',
                amount: 0,
                count: 1,
              },
            ]),
          },
          stageCostItems,
        ),
        createMonth(
          'may',
          {
            label: '5月',
            events: 6,
            salesMultiplier: stableUnitsPerEvent,
            onlineSalesFactor: 0,
            rehearsalCount: 4,
            rehearsalCost: 300,
            teacherCount: 4,
            teacherCost: 200,
            specialCosts: createStageCostValues(stageCostItems, [
              {
                itemId: 'stage-cost-material',
                amount: 6,
                count: 1,
              },
            ]),
          },
          stageCostItems,
        ),
        createMonth(
          'jun',
          {
            label: '6月',
            events: 6,
            salesMultiplier: stableUnitsPerEvent,
            onlineSalesFactor: 0,
            rehearsalCount: 4,
            rehearsalCost: 300,
            teacherCount: 4,
            teacherCost: 200,
            specialCosts: createStageCostValues(stageCostItems, [
              {
                itemId: 'stage-cost-material',
                amount: 6,
                count: 1,
              },
            ]),
          },
          stageCostItems,
        ),
        createMonth(
          'jul',
          {
            label: '7月',
            events: 6,
            salesMultiplier: stableUnitsPerEvent,
            onlineSalesFactor: 0,
            rehearsalCount: 4,
            rehearsalCost: 300,
            teacherCount: 4,
            teacherCost: 200,
            specialCosts: createStageCostValues(stageCostItems, [
              {
                itemId: 'stage-cost-material',
                amount: 6,
                count: 1,
              },
            ]),
          },
          stageCostItems,
        ),
        createMonth(
          'aug',
          {
            label: '8月',
            events: 6,
            salesMultiplier: stableUnitsPerEvent,
            onlineSalesFactor: 0,
            rehearsalCount: 4,
            rehearsalCost: 300,
            teacherCount: 4,
            teacherCost: 200,
            specialCosts: createStageCostValues(stageCostItems, [
              {
                itemId: 'stage-cost-material',
                amount: 6,
                count: 1,
              },
            ]),
          },
          stageCostItems,
        ),
      ],
      planning,
      'default',
      timelineTemplate,
      stageCostItems,
    ),
  }
}
